File: apps/api/services/gemini_service.py
"""
Gemini video analysis service with model fallback.
"""

import os
import base64
import time
import json
from typing import Any, List
from google import genai
from google.genai import types

# Configure Gemini API
api_key = os.getenv("GEMINI_API_KEY")

# Model fallback order
GEMINI_MODELS = [
    "gemini-2.0-flash",
    "gemini-2.5-flash-lite", 
    "gemini-2.0-flash-lite",
    "gemini-2.5-flash",
]

# Enhanced system prompt for pedagogical tutor
SYSTEM_PROMPT = """
Você é um tutor pedagógico especialista em criar experiências de aprendizado gamificadas e interativas.

SUA MISSÃO:
Analisar o vídeo educacional fornecido e criar desafios interativos que testem a compreensão do aluno em momentos estratégicos.

DIRETRIZES:
1. Identifique momentos-chave para pausar e verificar o conhecimento.
2. Crie quizzes de múltipla escolha e exercícios de código.
3. Distribua os desafios ao longo do vídeo.

SAÍDA ESPERADA (JSON):
{
  "challenges": [
    {
      "timestamp": 45,
      "timestampLabel": "00:45",
      "type": "quiz",
      "title": "Conceito de Variáveis",
      "content": "Qual a função principal de uma variável?",
      "options": ["Armazenar dados", "Loop", "Função", "Texto"],
      "correctAnswer": 0,
      "summary": "Explicação..."
    }
  ]
}
"""


def _call_gemini_with_fallback(client: genai.Client, contents: list, config: types.GenerateContentConfig) -> str:
    """
    Try multiple Gemini models with fallback on quota/not-found errors.
    """
    last_error = None
    
    for model in GEMINI_MODELS:
        try:
            print(f"Trying model: {model}")
            response = client.models.generate_content(
                model=model,
                contents=contents,
                config=config
            )
            if response.text:
                print(f"Success with model: {model}")
                return response.text
            else:
                print(f"Empty response from {model}, trying next...")
                continue
                
        except Exception as e:
            error_str = str(e)
            print(f"Model {model} failed: {error_str[:200]}")
            
            if "429" in error_str or "RESOURCE_EXHAUSTED" in error_str:
                last_error = e
                continue
            elif "404" in error_str or "NOT_FOUND" in error_str:
                last_error = e
                continue
            else:
                raise e
    
    raise Exception(f"All Gemini models failed. Last error: {last_error}")


async def analyze_video(video_base64: str, mime_type: str) -> List[dict[str, Any]]:
    """
    Analyze video using Google GenAI SDK with model fallback.
    """
    if not api_key:
        print("GEMINI_API_KEY not set, returning fallback")
        return _get_fallback_challenges()
    
    try:
        client = genai.Client(api_key=api_key)
        
        video_bytes = base64.b64decode(video_base64)
        
        contents = [
            types.Content(
                role="user",
                parts=[
                    types.Part.from_bytes(data=video_bytes, mime_type=mime_type),
                    types.Part.from_text(text=SYSTEM_PROMPT)
                ]
            )
        ]
        
        config = types.GenerateContentConfig(
            temperature=0.7,
            top_p=0.95,
            top_k=40,
            max_output_tokens=2048,
            response_mime_type="application/json"
        )
        
        response_text = _call_gemini_with_fallback(client, contents, config)
        data = json.loads(response_text)
        
        challenges = data.get('challenges', [])
        timestamp = int(time.time() * 1000)
        
        for i, challenge in enumerate(challenges):
            challenge['id'] = f"ai-{timestamp}-{i}"
            
        return challenges
            
    except Exception as e:
        print(f"Error analyzing video: {str(e)}")
        print("Falling back to mock challenges due to API error")
        return _get_fallback_challenges()


def _get_fallback_challenges() -> List[dict[str, Any]]:
    """Return mock challenges when AI fails."""
    return [
        {
            "id": "fallback-1",
            "timestamp": 30,
            "timestampLabel": "00:30",
            "type": "quiz",
            "title": "Conceito Inicial",
            "content": "Qual o objetivo principal apresentado?",
            "options": ["Opção A", "Opção B", "Opção C", "Opção D"],
            "correctAnswer": 0,
            "summary": "Fallback challenge generated due to AI error."
        }
    ]

File: services/geminiService.ts

import { GoogleGenAI, Type } from "@google/genai";
import { Challenge, ChallengeType } from "../types";

const ai = new GoogleGenAI({ apiKey: process.env.API_KEY || '' });

// Prompt de sistema para o tutor pedagógico
export const SYSTEM_PROMPT = `
Você é um tutor pedagógico especialista em tecnologia. Sua tarefa é analisar o vídeo anexado e identificar momentos chave onde um conceito técnico foi concluído.
Para cada momento, crie um desafio (Quiz ou Exercício de Código).

Regras de Saída:
1. Retorne APENAS um objeto JSON.
2. Identifique de 2 a 4 momentos no vídeo.
3. Use o formato de timestamp MM:SS para 'timestampLabel' e converta para segundos em 'timestamp'.
4. Para desafios de 'code', forneça uma 'task' clara e o 'correctAnswer' como o código esperado.
5. Para 'quiz', forneça 4 'options' e o 'correctAnswer' como o índice (0-3).
6. Inclua um 'summary' curto explicando o conceito que acabou de ser visto.

Dificuldade Adaptativa: Foque em conceitos fundamentais mas prepare a base para aumentar a complexidade.
`;

export async function analyzeVideo(videoBase64: string, mimeType: string): Promise<Challenge[]> {
  const model = ai.models.generateContent({
    model: 'gemini-1.5-flash',
    contents: {
      parts: [
        { inlineData: { data: videoBase64, mimeType } },
        { text: "Analise este vídeo e gere os desafios técnicos conforme as instruções de sistema." }
      ]
    },
    config: {
      systemInstruction: SYSTEM_PROMPT,
      responseMimeType: "application/json",
      responseSchema: {
        type: Type.OBJECT,
        properties: {
          challenges: {
            type: Type.ARRAY,
            items: {
              type: Type.OBJECT,
              properties: {
                timestamp: { type: Type.NUMBER },
                timestampLabel: { type: Type.STRING },
                type: { type: Type.STRING, enum: Object.values(ChallengeType) },
                title: { type: Type.STRING },
                content: { type: Type.STRING },
                options: {
                  type: Type.ARRAY,
                  items: { type: Type.STRING }
                },
                correctAnswer: { type: Type.STRING },
                summary: { type: Type.STRING }
              },
              required: ["timestamp", "type", "title", "content", "correctAnswer", "summary"]
            }
          }
        },
        required: ["challenges"]
      }
    }
  });

  const response = await model;
  const data = JSON.parse(response.text || '{"challenges": []}');

  // Add unique IDs to challenges
  return data.challenges.map((c: any, index: number) => ({
    ...c,
    id: `challenge-${index}-${Date.now()}`
  }));
}
